File: src/electron/store.ts
import { app, safeStorage } from "electron";
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import type { AccountConfig, ProviderId } from "../shared/types";
import type { FetchedItem } from "../providers/types";

// JSON persistence for account metadata, the app's own persistent item
// collection (upstream disappearance is a read signal, never a delete — see
// sync.ts), and preferences. Tokens are stored in the JSON too,
// safeStorage-encrypted (see the token storage note below).

/** Read items vanish this long after upstream last mentioned them. */
const RETAIN_READ_MS = 30 * 24 * 60 * 60 * 1000;
/** Hard cap on the collection; oldest (read first) beyond it are dropped. */
const ITEMS_CAP = 500;

/** A FetchedItem the app has adopted into its own collection. The transient
 * upstreamRead signal is consumed at upsert and never persisted. */
export interface StoredItem extends Omit<FetchedItem, "upstreamRead"> {
  read: boolean;
  /** ISO 8601 — when this item first entered the collection. */
  firstSeenAt: string;
  /** ISO 8601 — the last sync in which upstream still returned the item. */
  lastSeenUpstreamAt: string;
}

interface StoredAccount {
  /** Base64 of the safeStorage-encrypted token, or the plaintext token.
   * Absent when the token lives in the OS keychain instead. */
  token?: string;
  tokenEncrypted?: boolean;
  baseUrl?: string;
  username?: string;
}

interface StoreFile {
  accounts: Partial<Record<ProviderId, StoredAccount>>;
  items: Record<string, StoredItem>;
  /** Translucent background preference; absent → platform default. */
  glassEnabled?: boolean;
}

/** Pre-collection format, folded into `items` on first load. */
interface LegacyStoreFile {
  readIds?: string[];
  seenIds?: string[];
  cachedItems?: FetchedItem[];
}

export interface StoredAccountInfo {
  config: AccountConfig;
  username?: string;
}

export interface Store {
  getAccount(provider: ProviderId): StoredAccountInfo | undefined;
  setAccount(provider: ProviderId, config: AccountConfig, username: string): void;
  deleteAccount(provider: ProviderId): void;
  getItems(): StoredItem[];
  /** Adopts new items (unread) and refreshes known ones (fields updated,
   * lastSeenUpstreamAt bumped, local read preserved). Returns the ids that
   * were new to the collection. */
  upsertItems(items: FetchedItem[]): string[];
  markRead(ids: string[]): void;
  markAllRead(): void;
  /** Removes a disconnected provider's items immediately. */
  deleteProviderItems(provider: ProviderId): void;
  /** Retention: drops read items unseen upstream for 30 days, then enforces
   * the 500-item cap (oldest by updatedAt go first, read before unread). */
  prune(): void;
  /** undefined when the user never toggled it — use the platform default. */
  getGlassEnabled(): boolean | undefined;
  setGlassEnabled(enabled: boolean): void;
}

// --- Token storage -------------------------------------------------------
// Tokens live safeStorage-encrypted in the JSON store on every platform:
// Chromium's os_crypt derives the key from the macOS Keychain, Windows
// DPAPI, or the Linux Secret Service. A previous version kept tokens as OS
// keychain/keyring items instead; on macOS that re-prompted for access on
// every rebuild (item ACLs pin the exact build hash for apps without an
// Apple-issued certificate), so it was rolled back.

export function createStore(): Store {
  const file = join(app.getPath("userData"), "majordomo.json");

  let data: StoreFile = { accounts: {}, items: {} };
  let migrated = false;
  try {
    const parsed = JSON.parse(readFileSync(file, "utf8")) as Partial<StoreFile> & LegacyStoreFile;
    let items: Record<string, StoredItem> = {};
    if (parsed.items && typeof parsed.items === "object") {
      items = parsed.items;
    } else if (Array.isArray(parsed.cachedItems)) {
      // Legacy format: fold cachedItems + readIds into the collection.
      // seenIds is dropped — "new to the collection" replaces it.
      const readIds = new Set(Array.isArray(parsed.readIds) ? parsed.readIds : []);
      const now = new Date().toISOString();
      for (const item of parsed.cachedItems) {
        items[item.id] = { ...item, read: readIds.has(item.id), firstSeenAt: now, lastSeenUpstreamAt: now };
      }
      migrated = true;
      console.log(`majordomo: migrated ${parsed.cachedItems.length} cached items to the collection`);
    }
    data = {
      accounts: parsed.accounts && typeof parsed.accounts === "object" ? parsed.accounts : {},
      items,
      glassEnabled: typeof parsed.glassEnabled === "boolean" ? parsed.glassEnabled : undefined,
    };
  } catch {
    // Missing or corrupt file — start fresh.
  }

  function save(): void {
    try {
      mkdirSync(dirname(file), { recursive: true });
      // data holds only the current shape, so legacy keys (readIds, seenIds,
      // cachedItems) are dropped from the JSON by this write.
      writeFileSync(file, JSON.stringify(data));
    } catch (err) {
      console.error("majordomo: failed to save store:", err);
    }
  }

  function decodeToken(stored: StoredAccount): string | undefined {
    if (stored.token === undefined) return undefined;
    if (!stored.tokenEncrypted) return stored.token;
    try {
      return safeStorage.decryptString(Buffer.from(stored.token, "base64"));
    } catch {
      return undefined;
    }
  }

  function encodeToken(stored: StoredAccount, token: string): void {
    const encrypt = safeStorage.isEncryptionAvailable();
    stored.token = encrypt ? safeStorage.encryptString(token).toString("base64") : token;
    stored.tokenEncrypted = encrypt;
  }

  if (migrated) save();

  // Decoded tokens are memoized so decryption runs once per account per run.
  const tokenCache = new Map<ProviderId, string>();

  return {
    getAccount(provider) {
      const stored = data.accounts[provider];
      if (!stored) return undefined;
      let token = tokenCache.get(provider);
      if (token === undefined) {
        token = decodeToken(stored);
        if (token === undefined) return undefined;
        tokenCache.set(provider, token);
      }
      return { config: { token, baseUrl: stored.baseUrl }, username: stored.username };
    },

    setAccount(provider, config, username) {
      const account: StoredAccount = { baseUrl: config.baseUrl, username };
      encodeToken(account, config.token);
      tokenCache.set(provider, config.token);
      data.accounts[provider] = account;
      save();
    },

    deleteAccount(provider) {
      tokenCache.delete(provider);
      delete data.accounts[provider];
      save();
    },

    getItems() {
      return Object.values(data.items);
    },

    upsertItems(items) {
      if (items.length === 0) return [];
      const now = new Date().toISOString();
      const newIds: string[] = [];
      for (const { upstreamRead, ...item } of items) {
        // Local read state only ever moves toward read: the user reading it
        // here or upstream both count, and nothing un-reads an item.
        const existing = data.items[item.id];
        if (existing) {
          data.items[item.id] = {
            ...item,
            // Enriched fields can be absent on rounds where the provider
            // skipped the lookup — never forget what a prior sync learned.
            state: item.state ?? existing.state,
            author: item.author ?? existing.author,
            read: existing.read || upstreamRead === true,
            firstSeenAt: existing.firstSeenAt,
            lastSeenUpstreamAt: now,
          };
        } else {
          newIds.push(item.id);
          data.items[item.id] = {
            ...item,
            read: upstreamRead === true,
            firstSeenAt: now,
            lastSeenUpstreamAt: now,
          };
        }
      }
      save();
      return newIds;
    },

    markRead(ids) {
      let changed = false;
      for (const id of ids) {
        const item = data.items[id];
        if (item && !item.read) {
          item.read = true;
          changed = true;
        }
      }
      if (changed) save();
    },

    markAllRead() {
      let changed = false;
      for (const item of Object.values(data.items)) {
        if (!item.read) {
          item.read = true;
          changed = true;
        }
      }
      if (changed) save();
    },

    deleteProviderItems(provider) {
      let changed = false;
      for (const [id, item] of Object.entries(data.items)) {
        if (item.provider === provider) {
          delete data.items[id];
          changed = true;
        }
      }
      if (changed) save();
    },

    prune() {
      let changed = false;
      const cutoff = Date.now() - RETAIN_READ_MS;
      for (const [id, item] of Object.entries(data.items)) {
        if (item.read && Date.parse(item.lastSeenUpstreamAt) < cutoff) {
          delete data.items[id];
          changed = true;
        }
      }
      const all = Object.values(data.items);
      const excess = all.length - ITEMS_CAP;
      if (excess > 0) {
        all.sort((a, b) => {
          if (a.read !== b.read) return a.read ? -1 : 1;
          return a.updatedAt < b.updatedAt ? -1 : a.updatedAt > b.updatedAt ? 1 : 0;
        });
        for (const victim of all.slice(0, excess)) delete data.items[victim.id];
        changed = true;
      }
      if (changed) save();
    },

    getGlassEnabled() {
      return data.glassEnabled;
    },

    setGlassEnabled(enabled) {
      data.glassEnabled = enabled;
      save();
    },
  };
}

File: Sources/Majordomo/Store/Store.swift
// JSON persistence for accounts and the app's own persistent item collection
// (upstream disappearance is a read signal, never a delete — see SyncEngine).
// Tokens are stored in the JSON too, AES-GCM-encrypted (see TokenCipher).
// Accounts are keyed by AccountId, any number per provider.

import Foundation
import os

private let log = Logger(subsystem: "dev.majordomo.app", category: "store")

/// Read items vanish this long after upstream last mentioned them.
private let retainReadSeconds: TimeInterval = 30 * 24 * 60 * 60
/// Hard cap on the collection; oldest (read first) beyond it are dropped.
private let itemsCap = 500

/// A FetchedItem the app has adopted into its own collection. The transient
/// upstreamRead signal is consumed at upsert and never persisted.
struct StoredItem: Codable, Sendable {
  var id: String
  var accountId: AccountId
  var provider: ProviderId
  var kind: String
  var title: String
  var repo: String
  var url: String
  var reason: String
  var isMention: Bool
  var updatedAt: Date
  var state: String?
  var author: String?
  var body: String?
  var read: Bool
  /// When this item first entered the collection.
  var firstSeenAt: Date
  /// The last sync in which upstream still returned the item.
  var lastSeenUpstreamAt: Date

  var inboxItem: InboxItem {
    InboxItem(
      id: id, accountId: accountId, provider: provider, kind: kind,
      title: title, repo: repo, url: url, reason: reason, isMention: isMention,
      updatedAt: updatedAt, state: state, author: author, body: body, read: read
    )
  }
}

struct StoredAccount: Codable {
  var provider: ProviderId
  /// Base64 of the AES-GCM-sealed token, or the plaintext token.
  var token: String?
  var tokenEncrypted: Bool?
  var baseUrl: String?
  var username: String?
  var name: String?
  var avatarUrl: String?
  /// Orders the accounts list.
  var createdAt: Date?
}

struct StoreFile: Codable {
  var accounts: [String: StoredAccount] = [:]
  var items: [String: StoredItem] = [:]
}

struct StoredAccountInfo {
  let id: AccountId
  let provider: ProviderId
  /// nil when the stored token can't be read (Keychain denied) — the
  /// account still lists, but can't sync until reconnected.
  var config: AccountConfig?
  var username: String?
  var name: String?
  var avatarUrl: String?
  var baseUrl: String?
}

/// Not Sendable by design: owned and accessed exclusively by the SyncEngine
/// actor.
final class Store {
  private let fileURL: URL
  /// False for the in-memory fixture store (MAJORDOMO_DEMO) — nothing is
  /// ever written.
  private let persistent: Bool
  private var data = StoreFile()
  /// Lazy so a demo run (plaintext fixture tokens) never touches the
  /// Keychain at all.
  private lazy var key = TokenCipher.loadOrCreateKey()
  /// Decoded tokens, memoized so decryption runs once per account per run.
  private var tokenCache: [AccountId: String] = [:]

  /// ISO 8601 both ways; decoding tolerates fractional seconds.
  private static func decoder() -> JSONDecoder {
    let decoder = JSONDecoder()
    decoder.dateDecodingStrategy = .custom { decoder in
      let raw = try decoder.singleValueContainer().decode(String.self)
      guard let date = ISODate.parse(raw) else {
        throw DecodingError.dataCorrupted(.init(
          codingPath: decoder.codingPath,
          debugDescription: "unparseable date: \(raw)"
        ))
      }
      return date
    }
    return decoder
  }

  private static func encoder() -> JSONEncoder {
    let encoder = JSONEncoder()
    encoder.dateEncodingStrategy = .custom { date, encoder in
      var container = encoder.singleValueContainer()
      try container.encode(ISODate.format(date))
    }
    return encoder
  }

  /// An in-memory store seeded with fixtures; nothing persists.
  init(fixture: StoreFile) {
    fileURL = URL(fileURLWithPath: "/dev/null")
    persistent = false
    data = fixture
  }

  init(directory: URL? = nil) {
    persistent = true
    let base = directory ?? FileManager.default
      .urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
      .appending(path: "Majordomo")
    fileURL = base.appending(path: "majordomo.json")
    if let raw = try? Data(contentsOf: fileURL) {
      if let parsed = try? Self.decoder().decode(StoreFile.self, from: raw) {
        data = parsed
      } else {
        // An unreadable store must never be silently clobbered by the next
        // save — park a copy for recovery, then start fresh.
        let parked = fileURL.appendingPathExtension("corrupt")
        try? raw.write(to: parked, options: .atomic)
        log.error("store unreadable; original preserved as \(parked.lastPathComponent, privacy: .public)")
      }
    }
    // Missing file — start fresh.
  }

  private func save() {
    guard persistent else {
      return
    }
    do {
      try FileManager.default.createDirectory(
        at: fileURL.deletingLastPathComponent(),
        withIntermediateDirectories: true
      )
      try Self.encoder().encode(data).write(to: fileURL, options: .atomic)
    } catch {
      log.error("failed to save store: \(error)")
    }
  }

  private func decodeToken(_ stored: StoredAccount) -> String? {
    guard let token = stored.token else { return nil }
    guard stored.tokenEncrypted == true else { return token }
    guard let key else { return nil }
    return TokenCipher.decrypt(token, key: key)
  }

  // MARK: Accounts

  private func info(id: AccountId, _ stored: StoredAccount) -> StoredAccountInfo {
    var config: AccountConfig?
    if let cached = tokenCache[id] {
      config = AccountConfig(token: cached, baseUrl: stored.baseUrl)
    } else if let decoded = decodeToken(stored) {
      tokenCache[id] = decoded
      config = AccountConfig(token: decoded, baseUrl: stored.baseUrl)
    }
    return StoredAccountInfo(
      id: id,
      provider: stored.provider,
      config: config,
      username: stored.username,
      name: stored.name,
      avatarUrl: stored.avatarUrl,
      baseUrl: stored.baseUrl
    )
  }

  /// Every stored account, oldest first.
  func accounts() -> [StoredAccountInfo] {
    data.accounts
      .sorted { a, b in
        let left = a.value.createdAt ?? .distantPast
        let right = b.value.createdAt ?? .distantPast
        if left != right {
          return left < right
        }
        return a.key < b.key
      }
      .map { info(id: $0.key, $0.value) }
  }

  func getAccount(_ id: AccountId) -> StoredAccountInfo? {
    guard let stored = data.accounts[id] else { return nil }
    return info(id: id, stored)
  }

  func setAccount(
    _ id: AccountId, provider: ProviderId, config: AccountConfig, profile: ProviderProfile
  ) {
    var account = StoredAccount(
      provider: provider,
      baseUrl: config.baseUrl,
      username: profile.username,
      name: profile.name,
      avatarUrl: profile.avatarUrl,
      createdAt: data.accounts[id]?.createdAt ?? Date()
    )
    if let key, let sealed = TokenCipher.encrypt(config.token, key: key) {
      account.token = sealed
      account.tokenEncrypted = true
    } else {
      account.token = config.token
      account.tokenEncrypted = false
    }
    tokenCache[id] = config.token
    data.accounts[id] = account
    save()
  }

  /// Refreshes the owner's profile fields without touching the token.
  func updateProfile(_ id: AccountId, profile: ProviderProfile) {
    guard var account = data.accounts[id] else {
      return
    }
    account.username = profile.username
    account.name = profile.name
    account.avatarUrl = profile.avatarUrl
    data.accounts[id] = account
    save()
  }

  func deleteAccount(_ id: AccountId) {
    tokenCache.removeValue(forKey: id)
    data.accounts.removeValue(forKey: id)
    save()
  }

  // MARK: Items

  func getItems() -> [StoredItem] {
    Array(data.items.values)
  }

  /// Adopts new items (unread) and refreshes known ones (fields updated,
  /// lastSeenUpstreamAt bumped, local read preserved). Item ids compose as
  /// "\(accountId):\(externalId)". Returns the ids new to the collection.
  func upsertItems(
    _ items: [FetchedItem], accountId: AccountId, provider: ProviderId
  ) -> [String] {
    guard !items.isEmpty else { return [] }
    let now = Date()
    var newIds: [String] = []
    for fetched in items {
      let id = "\(accountId):\(fetched.externalId)"
      let existing = data.items[id]
      if existing == nil {
        newIds.append(id)
      }
      data.items[id] = StoredItem(
        id: id,
        accountId: accountId,
        provider: provider,
        kind: fetched.kind,
        title: fetched.title,
        repo: fetched.repo,
        url: fetched.url,
        reason: fetched.reason,
        isMention: fetched.isMention,
        updatedAt: fetched.updatedAt,
        // Enriched fields can be absent on rounds where the provider
        // skipped the lookup — never forget what a prior sync learned.
        state: fetched.state ?? existing?.state,
        author: fetched.author ?? existing?.author,
        body: fetched.body ?? existing?.body,
        // Local read state only ever moves toward read: the user reading it
        // here or upstream both count, and nothing un-reads an item.
        read: (existing?.read ?? false) || fetched.upstreamRead == true,
        firstSeenAt: existing?.firstSeenAt ?? now,
        lastSeenUpstreamAt: now
      )
    }
    save()
    return newIds
  }

  func markRead(_ ids: [String]) {
    var changed = false
    for id in ids {
      if let item = data.items[id], !item.read {
        data.items[id]?.read = true
        changed = true
      }
    }
    if changed {
      save()
    }
  }

  func markAllRead() {
    var changed = false
    for (id, item) in data.items where !item.read {
      data.items[id]?.read = true
      changed = true
    }
    if changed {
      save()
    }
  }

  /// Removes a disconnected account's items immediately.
  func deleteAccountItems(_ accountId: AccountId) {
    let before = data.items.count
    data.items = data.items.filter { $0.value.accountId != accountId }
    if data.items.count != before {
      save()
    }
  }

  /// Retention: drops read items unseen upstream for 30 days, then enforces
  /// the 500-item cap (oldest by updatedAt go first, read before unread).
  func prune() {
    var changed = false
    let cutoff = Date().addingTimeInterval(-retainReadSeconds)
    for (id, item) in data.items where item.read && item.lastSeenUpstreamAt < cutoff {
      data.items.removeValue(forKey: id)
      changed = true
    }
    let excess = data.items.count - itemsCap
    if excess > 0 {
      let victims = data.items.values
        .sorted { a, b in
          if a.read != b.read {
            return a.read
          }
          return a.updatedAt < b.updatedAt
        }
        .prefix(excess)
      for victim in victims {
        data.items.removeValue(forKey: victim.id)
      }
      changed = true
    }
    if changed {
      save()
    }
  }
}
